lib/pure/math.nim

```diff
@@ -158,6 +158,23 @@ proc classify*(x: float): FloatClass =
     return fcSubnormal
   return fcNormal
 
+proc approxEqual*[T: SomeFloat](x, y: T; ulp: Natural = 4): bool {.noSideEffect.} =
+  ## Checks if two float values are considered equal, using machine epsilon.
```

Comment:
`4` is subjective, individual tests can define their own tolerance and use syntax sugar operator like `=~` eg:
template `=~`(a, b): untyped = almostEqual(a, b, 5)

Reply:
`doAssert 1.0000 =~ 0.99999`

This check doesn't work, I tried `ulp` values up to 2 billion.

Reply:
Found `kMaxUlps = 4` [googletest](https://github.com/google/googletest/blob/3c95bf552405fd0cc63cea0ca2f6c4cd89c8d356/googletest/include/gtest/internal/gtest-internal.h#L290)

Reply:
That googletest comment is incorrect.  On recent (post 2011-ish) Intel/AMD CPUs with recent gcc/clang backends, much (most?) FP calcs are done in SIMD registers not the 80-bit x87 FP stack machine.  SIMD are only done to IEEE double precision.  Precision vagueness (16, 32, 64, 80, 128?) are another reason that counting matching digits rather than mismatching [is maybe more convenient](https://github.com/nim-lang/RFCs/issues/275#issuecomment-716784434).  Counting matching digits from the front (or number-units like 1e-9 rather than bit units) matches across precision (to the extent it makes sense).  E.g., 20 matching bits means roughly the same thing for all precisions >= 20 mantissa bits.

All this said, I think 2..4 bits is an ok ULP default.  It doesn't take much calculation to go from the guarantees of 0.5 for [+-*/] of elemental ops up to 4 bits of accumulated error.  A lot of C libs did not used to calculate transcendental functions all the way to the last place, but I think that has gotten better in recent decades.

Reply:
> It's a proc parameter with a default value so it can still be changed.

once you set it as an optional, changing it is a (however mild) breaking change; and you can't make it required;
whereas if you make it required, you can later make it optional without a breaking change.